server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -468,12 +475,20 @@ public void updateShardState(final ShardRouting newRouting,
                     if (resyncStarted == false) {
                         throw new IllegalStateException("cannot start resync while it's already in progress");
                     }
-                    indexShardOperationPermits.asyncBlockOperations(
-                        30,
-                        TimeUnit.MINUTES,
+                    bumpPrimaryTerm(newPrimaryTerm,
                         () -> {
                             shardStateUpdated.await();
+                            assert pendingPrimaryTerm == newPrimaryTerm :
+                                "shard term changed on primary. expected [" + newPrimaryTerm + "] but was [" + pendingPrimaryTerm + "]" +
+                                ", current routing: " + currentRouting + ", new routing: " + newRouting;
+                            assert operationPrimaryTerm < newPrimaryTerm;
                             try {
+                                synchronized (mutex) {
+                                    assert shardRouting.primary();
+                                    // do these updates under the mutex as this otherwise races with subsequent calls of updateShardState
```

Comment:
what race conditions do refer to?

Reply:
I think this is obsolete and does not need to be done under the mutex anymore. It came from an earlier iteration where I had not introduced the `relocated` state in ReplicationTracker yet, and where I was using both `replicationTracker.isPrimaryMode()` and comparing `operationPrimaryTerm` to `pendingPrimaryTerm` to figure out if a shard had possibly relocated and wanted this to be an atomic thing.

Reply:
One problematic place I see is the assertion at the end of updateShardState which checks both isPrimaryMode and the operationTerm + pendingPrimaryTerm. If we don't update the replication tracker + the operationPrimaryTerm atomically under the mutex, this invariant might be violated.